### PR TITLE
fix problem with loop optimization (#742)

### PR DESCRIPTION
**Pull request by dsyme**
Provides a modified and reviewed fix for the problem reported here: https://github.com/Microsoft/visualfsharp/pull/742

I've made the code cleaner and added additional constraints to ensure the optimization only applies to the compiled forms of ``for ... in  ... do ... `` loops

### CODE REVIEW

**Comment by robertpi** [comment]
Cool. Looks good. I did wonder if checking the value in the enumerator and previous let binding was possible, but thought what I had was good enough. Would never have thought to make the change in the type checker either.

File: tests/fsharp/core/seq/test.fsx [hunk 1]
```diff
@@ -487,6 +487,31 @@ check "hfhdfsjkfur34"
         Failure "ss!!!" -> results := "caught"::!results
     !results)
     ["caught";"ssDispose";"eDispose"]
+
+// Check https://github.com/Microsoft/visualfsharp/pull/742
```
Comment (brodyberg) [inline tests/fsharp/core/seq/test.fsx]: In lieu of GitHub not lasting forever, one line of context would be helpful

**Comment by KevinRansom** [comment]
Thanks for fixing this merged.

**Comment by dsyme** [comment]
A not-so-pretty workaround for this issue is here: https://github.com/Microsoft/visualfsharp/issues/759#issuecomment-162243299.  (note however it requires changes to user code)

File: src/fsharp/TastOps.fs [hunk 2]
```diff
@@ -7748,86 +7748,120 @@ let (|RangeInt32Step|_|) g expr =
...
+    | Let (enumerableVar, enumerableExpr, _,
+           Let (enumeratorVar, GetEnumeratorCall enumerableVar2, enumeratorBind,
+              TryFinally (WhileLoopForCompiledForEachExpr (_, Let (elemVar,_,_,bodyExpr), _), _))) 
+                 // Apply correctness conditions to ensure this really is a compiled for-each expression.
+                 when valRefEq g (mkLocalValRef enumerableVar) enumerableVar2 &&
+                      enumerableVar.IsCompilerGenerated &&
+                      enumeratorVar.IsCompilerGenerated &&
+                      let fvs = (freeInExpr CollectLocals bodyExpr)
+                      not (Zset.contains enumerableVar fvs.FreeLocals) && 
+                      not (Zset.contains enumeratorVar fvs.FreeLocals) ->
```
Comment (KevinRansom) [inline src/fsharp/TastOps.fs]: Don is this right?

Comment (dsyme) [inline src/fsharp/TastOps.fs]: Yes, I think so - is there something I've missed?  The "bodyExpr" is the body of the while loop, but it shouldn't directly refer to either enumerableVar (the variable holding the enumerable collection, which is consumed in GetEnumeratorCall) or enumeratorVar (the variable holding the enumerator for the collection, which is consumed in the binding for the elemVar "let").

**Comment by KevinRansom** [comment]
On the internal build this fails with:

The OSS build succeeds though.

**Comment by mexx** [comment]
fails for me also in the wild ;) already on the `FSharp.Compiler-proto.dll`

**Comment by enricosada** [comment]
why this difference with appveyor build?

**Comment by mexx** [comment]
I think it's a matter of what `fsc` version is used to build the `proto` compiler, will check which version was used in my setup.

**Comment by mexx** [comment]
Here is the start of the failing call on my setup
```..\..\..\lkg\FSharp-4.0.30319.1\bin\fsc.exe -o:obj\Proto\net40\FSharp.Build-proto.dll ...```
on AppVeyor
```C:\Program Files (x86)\Microsoft SDKs\F#\3.1\Framework\v4.0\fsc.exe -o:obj\Proto\net40\FSharp.Build-proto.dll```

**Comment by KevinRansom** [comment]
the appveyor build uses F# 3.1, that is the difference between build success and failure.

**Comment by mexx** [comment]
What are the reasons why AppVeyor build doesn't use the LKG version, like the local ones?
Shouldn't the build take the same path as local ones?

**Comment by KevinRansom** [comment]
It'seems a bug in the script, we are looking at fixing it.

**Comment by KevinRansom** [comment]
pulled

**Pull request closed without merging**